neqo-transport/src/connection.rs

```diff
@@ -3223,4 +3249,214 @@ mod tests {
+                assert!(matches!(pkt, Output::Callback(_)));
+                break;
+            }
+        }
+
+        assert_eq!(c_tx_dgrams.len(), 11);
+
+        // Recv all pkts and ACK
+        qinfo!("server");
+        for c_dgram in c_tx_dgrams {
+            server.process_input(c_dgram, now)
+        }
+        assert_eq!(server.stream_recv(2, &mut srv_buf).unwrap(), (12283, false));
+        let mut s_tx_dgrams = Vec::new();
+        while let Output::Datagram(dg) = server.process_output(now) {
+            s_tx_dgrams.push(dg);
+        }
+
+        // ACK pkts 0-10
+        assert_eq!(s_tx_dgrams.len(), 1);
+
+        // Handle ack, send more
+        qinfo!("client");
+        for s_dgram in s_tx_dgrams {
+            client.process_input(s_dgram, now)
+        }
+
+        assert_eq!(client.stream_send(2, &[0xab; 100_000]).unwrap(), 49152);
+
+        let mut c_tx_dgrams = Vec::new();
+        while let Output::Datagram(dg) = client.process_output(now) {
+            c_tx_dgrams.push(dg);
+        }
+
+        assert_eq!(c_tx_dgrams.len(), 21);
+
+        // Recv all pkts and ACK
+        qinfo!("server");
+        for c_dgram in c_tx_dgrams {
+            server.process_input(c_dgram, now)
+        }
+        assert_eq!(server.stream_recv(2, &mut srv_buf).unwrap(), (24617, false));
+        let mut s_tx_dgrams = Vec::new();
+        while let Output::Datagram(dg) = server.process_output(now) {
+            s_tx_dgrams.push(dg);
+        }
+
+        // ACK pkts 0-31
+        assert_eq!(s_tx_dgrams.len(), 1);
+
+        // Handle ack, send more
+        qinfo!("client");
+        for s_dgram in s_tx_dgrams {
+            client.process_input(s_dgram, now)
+        }
+
+        // Cram more into the stream
+        assert_eq!(client.stream_send(2, &[0xab; 100_000]).unwrap(), 36900);
+
+        let mut c_tx_dgrams = Vec::new();
+        while let Output::Datagram(dg) = client.process_output(now) {
+            c_tx_dgrams.push(dg);
+        }
+
+        assert_eq!(c_tx_dgrams.len(), 41);
+
+        // Packets lost!
+        qinfo!("server");
+        for _ in 0..25 {
+            c_tx_dgrams.remove(10);
+        }
+        for c_dgram in c_tx_dgrams {
+            server.process_input(c_dgram, now)
+        }
+        assert_eq!(server.stream_recv(2, &mut srv_buf).unwrap(), (12005, false));
+        let mut s_tx_dgrams = Vec::new();
+        while let Output::Datagram(dg) = server.process_output(now) {
+            s_tx_dgrams.push(dg);
+        }
+        assert_eq!(s_tx_dgrams.len(), 1);
+
+        // Handle ack, transition slow start -> cong avoidance due to dropped
+        // pkts
+        qinfo!("client");
+        for s_dgram in s_tx_dgrams {
+            client.process_input(s_dgram, now)
+        }
+        let cong_cwnd = client.loss_recovery.cwnd();
+        assert_eq!(cong_cwnd, 35116);
+
+        // Cram more into the stream, but nothing added
+        assert_eq!(client.stream_send(2, &[0xac; 100_000]).unwrap(), 0);
```

Comment:
There is a function you can extract from here, it's called `send_on_stream_2_and_check_that_the_client_sends_X_packets_that_add_to_cwnd_Y()` or `send_and_check_cwnd()`.  That can return the datagrams that were produced.

Reply:
Good idea. Currently called `client_send_bytes()`. I really don't like writing too much helper code for tests but this does seem handy.